sources/packages/api/src/services/designation-agreement/designation-agreement.service.ts

```diff
@@ -0,0 +1,99 @@
+import { Connection } from "typeorm";
+import {
+  DesignationAgreement,
+  DesignationAgreementLocation,
+  DesignationAgreementStatus,
+  Institution,
+  InstitutionLocation,
+  User,
+} from "../../database/entities";
+
+/**
+ * Manages the operations needed for designation agreements that are submitted by the institutions
+ * and then need to be reviewed by the Ministry for approval or denial. While the Ministry is
+ * approving the designation it can also edit dates and locations as needed.
+ */
+@Injectable()
+export class DesignationAgreementService extends RecordDataModelService<DesignationAgreement> {
+  constructor(connection: Connection) {
+    super(connection.getRepository(DesignationAgreement));
+  }
+
+  /**
+   * Initiates a new designation agreement request. This action
+   * is meant to me initiated by the institution signing officer
+   * for further assessment of the Ministry.
+   * The designation agreement and the location that are part of it will
+   * be saved as part of the same DB transaction.
+   * @param institutionId institution id requesting the designation.
+   * @param submittedData dynamic data that represents the designation requested.
+   * @param submittedByUserId institution user submitting the designation.
+   * @param submittedDate date that the institution user submitted the designation.
+   * @param requestedLocationsIds ids of the locations requested to be designated.
+   * @returns the new designation agreement created.
+   */
+  async submitDesignationAgreement(
+    institutionId: number,
+    submittedData: any,
+    submittedByUserId: number,
+    submittedDate: Date,
+    requestedLocationsIds: number[],
+  ): Promise<DesignationAgreement> {
+    const submittedByUser = { id: submittedByUserId } as User;
+    const newDesignation = new DesignationAgreement();
+    newDesignation.institution = { id: institutionId } as Institution;
+    newDesignation.submittedData = submittedData;
+    newDesignation.designationStatus = DesignationAgreementStatus.Pending;
+    newDesignation.submittedBy = submittedByUser;
+    newDesignation.submittedDate = submittedDate;
+    newDesignation.creator = submittedByUser;
+    newDesignation.designationAgreementLocations = requestedLocationsIds.map(
+      (locationId: number) => {
+        const newLocation = new DesignationAgreementLocation();
+        newLocation.institutionLocation = {
+          id: locationId,
+        } as InstitutionLocation;
+        newLocation.requested = true;
+        newLocation.creator = submittedByUser;
+        return newLocation;
+      },
+    );
+    return this.repo.save(newDesignation);
+  }
+
+  /**
+   * Retrieve the designation agreement information and
+   * the associated locations approvals.
+   * @param designationId designation id.
+   * @param institutionId institution id.
+   * @returns designation agreement information.
+   */
+  async getDesignationById(
```

Comment:
IMO,  here designation results are dependent upon both `designationId` and `institutionId`, I think it will be to rename the function. because when I see `getDesignationById`, I think it should search only with `designationId`

Reply:
The institution id is included for authorization purpose only. The method is really to get the designation by its Id.
I would avoid renaming it to something like getDesignationByIdAndInstitution.
Maybe we can call it getInstitutionDesignationById? Any thoughts?

Reply:
ya `getInstitutionDesignationById` make sense to me

Reply:
I agree. We are using the same way in many place like applicationId and studentId|userId .